### ISSUE TITLE
Timeouts with router.ipc

### ISSUE THREAD

**Issue by tk-hendrik** [csirtgadgets/bearded-avenger#204]
Hi,

Question on top: Where to start debugging?

I ran into timeouts during consuming larger sets of data.
csirtg-smrt starts normal, downloading dataplane ssh-client feed and parsing it.
After a few secons (and many many logs from router & httpd) csirtg-smrt will die with its last words:

```
Dez 21 11:06:14 localhorst cif-httpd[62501]: 2016-12-21 11:06:14,802 - WARNING - cifsdk.client.zeromq[100][Thread-1512] - timeout trying to receive, retrying...
Dez 21 11:07:44 localhorst cif-httpd[62501]: 2016-12-21 11:07:44,893 - WARNING - cifsdk.client.zeromq[100][Thread-1512] - timeout trying to receive, retrying...
Dez 21 11:09:14 localhorst cif-httpd[62501]: 2016-12-21 11:09:14,983 - WARNING - cifsdk.client.zeromq[100][Thread-1512] - timeout trying to receive, retrying...
Dez 21 11:10:45 localhorst cif-httpd[62501]: 2016-12-21 11:10:45,073 - WARNING - cifsdk.client.zeromq[100][Thread-1512] - timeout trying to receive, retrying...
```
Systemd detects the die and respawn... infinite loop.

More Logs:
```
Dez 21 10:56:40 localhorst csirtg-smrt[66312]: 2016-12-21 10:56:40,731 - DEBUG - cifsdk.client.http[171] - http://localhost:5000/indicators
Dez 21 10:56:40 localhorst cif-router[62468]: 2016-12-21 10:56:40,747 - INFO - elasticsearch[66][MainThread] - GET http://localhost:9200/tokens/token/_search [status:200 request:0.001s]
Dez 21 10:56:44 localhorst cif-router[62468]: 2016-12-21 10:56:44,753 - INFO - elasticsearch[66][MainThread] - HEAD http://localhost:9200/indicators-2016.12 [status:200 request:0.001s]
Dez 21 10:56:44 localhorst cif-router[62468]: 2016-12-21 10:56:44,758 - INFO - elasticsearch[66][MainThread] - GET http://localhost:9200/indicators-*/indicator/_search?timeout=120s&size=5000 [status:200 request:0.003s]
Dez 21 10:58:10 localhorst cif-httpd[62501]: 2016-12-21 10:58:10,826 - WARNING - cifsdk.client.zeromq[100][Thread-1128] - timeout trying to receive, retrying...
Dez 21 10:59:40 localhorst cif-httpd[62501]: 2016-12-21 10:59:40,916 - WARNING - cifsdk.client.zeromq[100][Thread-1128] - timeout trying to receive, retrying...
Dez 21 11:01:11 localhorst cif-httpd[62501]: 2016-12-21 11:01:11,007 - WARNING - cifsdk.client.zeromq[100][Thread-1128] - timeout trying to receive, retrying...
Dez 21 11:02:41 localhorst cif-httpd[62501]: 2016-12-21 11:02:41,097 - WARNING - cifsdk.client.zeromq[100][Thread-1128] - timeout trying to receive, retrying...
Dez 21 11:04:11 localhorst cif-httpd[62501]: 2016-12-21 11:04:11,187 - WARNING - cifsdk.client.zeromq[100][Thread-1128] - timeout trying to receive, retrying...
Dez 21 11:04:11 localhorst cif-httpd[62501]: 2016-12-21 11:04:11,187 - ERROR - cif.httpd[366][Thread-1128] - timeout waiting for: ipc:///var/lib/cif/router.ipc
Dez 21 11:04:11 localhorst cif-httpd[62501]: 2016-12-21 11:04:11,188 - INFO - werkzeug[87][Thread-1128] - 127.0.0.1 - - [21/Dec/2016 11:04:11] "POST /indicators HTTP/1.1" 408 -
Dez 21 11:04:11 localhorst csirtg-smrt[66312]: 2016-12-21 11:04:11,189 - DEBUG - cifsdk.client.http[88] - request failed: 408
Dez 21 11:04:11 localhorst csirtg-smrt[66312]: Traceback (most recent call last):
Dez 21 11:04:11 localhorst csirtg-smrt[66312]:   File "/etc/cif/cif-venv/lib/python2.7/site-packages/csirtg_smrt-0.0.0a16+1.g7a01940-py2.7.egg/csirtg_smrt/smrt.py", line 325, in main
Dez 21 11:04:11 localhorst csirtg-smrt[66312]:     for i in s.process(r, f, limit=args.limit, data=data, filters=filters):
Dez 21 11:04:11 localhorst csirtg-smrt[66312]:   File "/etc/cif/cif-venv/lib/python2.7/site-packages/csirtg_smrt-0.0.0a16+1.g7a01940-py2.7.egg/csirtg_smrt/smrt.py", line 187, in process
Dez 21 11:04:11 localhorst csirtg-smrt[66312]:     self.send_indicators(indicator_batch)
Dez 21 11:04:11 localhorst csirtg-smrt[66312]:   File "/etc/cif/cif-venv/lib/python2.7/site-packages/csirtg_smrt-0.0.0a16+1.g7a01940-py2.7.egg/csirtg_smrt/smrt.py", line 166, in send_indicators
Dez 21 11:04:11 localhorst csirtg-smrt[66312]:     self.client.indicators_create(i)
Dez 21 11:04:11 localhorst csirtg-smrt[66312]:   File "build/bdist.linux-x86_64/egg/cifsdk/client/http.py", line 172, in indicators_create
Dez 21 11:04:11 localhorst csirtg-smrt[66312]:     rv = self._post(uri, data)
Dez 21 11:04:11 localhorst csirtg-smrt[66312]:   File "build/bdist.linux-x86_64/egg/cifsdk/client/http.py", line 97, in _post
Dez 21 11:04:11 localhorst csirtg-smrt[66312]:     raise TimeoutError('timeout')
Dez 21 11:04:11 localhorst csirtg-smrt[66312]: TimeoutError: timeout
Dez 21 11:04:11 localhorst csirtg-smrt[66312]: Traceback (most recent call last):
Dez 21 11:04:11 localhorst csirtg-smrt[66312]:   File "/etc/cif/cif-venv/bin/csirtg-smrt", line 11, in <module>
Dez 21 11:04:11 localhorst csirtg-smrt[66312]:     load_entry_point('csirtg-smrt==0.0.0a16+1.g7a01940', 'console_scripts', 'csirtg-smrt')()
Dez 21 11:04:11 localhorst csirtg-smrt[66312]:   File "/etc/cif/cif-venv/lib/python2.7/site-packages/csirtg_smrt-0.0.0a16+1.g7a01940-py2.7.egg/csirtg_smrt/smrt.py", line 350, in main
Dez 21 11:04:11 localhorst csirtg-smrt[66312]:     raise e
Dez 21 11:04:11 localhorst csirtg-smrt[66312]: cifsdk.exceptions.TimeoutError: timeout
Dez 21 11:04:11 localhorst systemd[1]: csirtg-smrt.service: Main process exited, code=exited, status=1/FAILURE
Dez 21 11:04:11 localhorst systemd[1]: csirtg-smrt.service: Unit entered failed state.
Dez 21 11:04:11 localhorst systemd[1]: csirtg-smrt.service: Failed with result 'exit-code'.
Dez 21 11:04:11 localhorst systemd[1]: csirtg-smrt.service: Service hold-off time over, scheduling restart.
Dez 21 11:04:11 localhorst systemd[1]: Stopped csirtg-smrt service.
Dez 21 11:04:11 localhorst systemd[1]: Started csirtg-smrt service.
Dez 21 11:04:11 localhorst csirtg-smrt[68089]: 2016-12-21 11:04:11,819 - INFO - root[271] - loglevel is: DEBUG
Dez 21 11:04:11 localhorst csirtg-smrt[68089]: 2016-12-21 11:04:11,825 - DEBUG - csirtg_smrt.archiver[79] - database path: sqlite:////var/lib/cif/smrt.db
Dez 21 11:04:11 localhorst csirtg-smrt[68089]: 2016-12-21 11:04:11,825 - INFO - root[295] - random delay is 0, then running every 60 min after that
Dez 21 11:04:11 localhorst csirtg-smrt[68089]: 2016-12-21 11:04:11,825 - INFO - root[311] - starting...
```

**Comment by wesyoung** [comment]
what release are you on/and/or are you debugging with, master? @JustinAzoff and i have made some *significant* improvements to the core that may solve some of this... but admittedly we're just circling around to the elasticsearch part, so that may be a blocker still (can you test your feeds with SQLite?)

i'm trying to cut a release this week that'll incorporate this at an sqlite level, then use that as the example in cleaning up the ES functionality where things are probably slower cause we're storing stuff over their transport, etc...

**Comment by JustinAzoff** [comment]
yeah.. if it is using smrt in fireball mode, but ES store can't upsert the 500 records fast enough it will time out.

I ran into that when we first got fireball mode working again.. sqlite store was slow, but since it had only being doing 1 upsert at a time, it never timed out.  As soon as smrt started sending cif 500 indicators with each upsert taking 500ms, that meant it would take 250s to do the upsert, so it would timeout every time.

Initially I worked around the issue by having smrt use a fireball batch size of only 20 until I could optimize the upsert. (now it does 500 in about a second)

**Comment by tk-hendrik** [comment]
> what release are you on/and/or are you debugging with, master?
No fun without master :) - Means yes. Just checked out this morning, shortly before opening this issue.

> yeah.. if it is using smrt in fireball mode, but ES store can't upsert the 500 records fast enough it will time out.
smrt is currently running without fireball mode. May the limit param helps here?

**Comment by tk-hendrik** [comment]
> i'm trying to cut a release this week that'll incorporate this at an sqlite level, then use that as the example in cleaning up the ES functionality where things are probably slower cause we're storing stuff over their transport, etc...

As soon as I can help out, just drop me some interesting commits to adapt them to ES

**Comment by JustinAzoff** [comment]
ah.. if it's not using fireball the single item updates are probably causing the store to buffer them after a while so it can do one batch upsert which could mean it is doing even more than 500..  

Turning on fireball mode and setting the size to 2 may actually make it work for now as a workaround.  (since the store just uses whatever batch size you give it).

The problem looks like indicators_upsert in elasticsearch store doesn't do much batching.

I don't know much about ES, but I bet moving

                     es.indices.flush(index=_current_index())

to the end of the function so it is only called once would give a huge perf win.

Basically, indicators_upsert is going to be called with a `data` list of ~500 indicators, and the function should do the least amount of round trips and commits/flushes/whatever as possible.

**Comment by wesyoung** [comment]
flush was "outside" had to move it up based on how the upsert of "rolling into a more recent index" works[ed]. that was before fireball really worked, or our other improvements, but i have a feeling that's going to be tricky and may require a different approach to handling an index [so i wouldn't move it out quite yet... it will lead to race conditions... esp as you start multi-threading out the instances sideways...].

that one needs some thought.

**Comment by JustinAzoff** [comment]
The docs seem to indicate that it is optional anyway:

https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-flush.html

**Comment by wesyoung** [comment]
it is, if you don't call it it happens (i think?) around once per sec or something... the issue becomes, if we're gathering / upserting from / to multiple ES nodes you start running into concurrency and other race issues.. which is why i think i need to re-look at it from another angle now that we know more about how the data is flowing front to back...

**Comment by tk-hendrik** [comment]
Hi,
please hit me next time if I'll not provide DEBUG Logs.
Yesterday I thought it may be a problem just with the last message since `--limit 20` shows the same behaviour. Couldn't be sizing problem on the socket or similar.
Today I started router with debug and guess what: `cif.gatherer.asn` is working like a horse for the enrichment and parallel `cif-httpd` throws those timeouts.

Just a guess but this might not be an only elasticsearch store 'problem'.

**Comment by wesyoung** [comment]
that's fair :)

in the next release, [and this should be in master] we actually disabled the peer lookup by default to help with this. the other question is, are you using a local bind cache as described by the playbooks? 

@JustinAzoff has another tool that we're testing the background that enables you to do that asn lookup stuff "in house" and in bulk:

https://github.com/JustinAzoff/asnlookup

just haven't had a chance to work on integrating it yet.

i'm guessing the feed you're running up against this on is a very large IP feed.. and we've been mostly testing with very large DGA feeds... so we may need to work that more closely into our testing too...

**Comment by JustinAzoff** [comment]
That shouldn't cause timeouts though.  I still think it would be indicators_upsert working through a large batch and causing things to time out.

I would apply this patch to get the ES store to log when it starts/finishes doing a batch, and what the size of the batch was.
```

diff --git a/cif/store/zelasticsearch/indicator.py b/cif/store/zelasticsearch/indicator.py
index 1e786fc..2c41a8c 100644
--- a/cif/store/zelasticsearch/indicator.py
+++ b/cif/store/zelasticsearch/indicator.py
@@ -15,6 +15,7 @@ import re
 from base64 import b64decode, b64encode
 import binascii
 import socket
+import time
 
 VALID_FILTERS = ['indicator', 'confidence', 'provider', 'itype', 'group', 'tags']
 
@@ -137,6 +138,9 @@ class IndicatorMixin(object):
         if isinstance(data, dict):
             data = [data]
 
+        start_time = time.time()
+        self.logger.info('Upserting %d indicators..', len(data))
+
         n = 0
         tmp = {}  # to deal with es flushing
 
@@ -212,6 +216,7 @@ class IndicatorMixin(object):
                     n += 1
                     tmp[d['indicator']].add(d['lasttime'])
 
+        self.logger.info('Upserting %d indicators.. took %0.2f seconds', len(data), time.time() - start_time)
         return n
 
     def indicators_search(self, filters, sort=None, raw=False):

```

**Comment by wesyoung** [comment]
i'm gonna merge this under #116. cutting a mid-stream release today, then circling around to re-factoring ES a bit with the performance things we learned in sqlite and hopefully get that plus better logging plus better sqlite in the next release.

**Comment by wesyoung** [comment]
@tk-hendrik i'd be interested if you have the same problem too by manually turning off the asn gatherer.. as an aside, @JustinAzoff  and i are putting together a 1m ip feed to factor into our QA testing env's too which should help us spot these on our end more clearly..

need to fiddle a bit with awscodebuild/deploy and automate more of the QA stuff (outside of what travis can do)... but that's a diff topic.. :)

**Comment by tk-hendrik** [comment]
Back from holidays today...

>  the other question is, are you using a local bind cache as described by the playbooks?

It was installed and configured but not used by resolv.conf. Since my system is behind some measurements I have to adjust bind to forward all to our company dns - no direct outbound allowed here. As long the local bind wouldn't increase the cache timer I guess this wouldn't help much for cif.

>  i'd be interested if you have the same problem too by manually turning off the asn gatherer.. 

After disabling `asn.py*` no more timeout error.

**Comment by tk-hendrik** [comment]
Back from holidays today...

>  the other question is, are you using a local bind cache as described by the playbooks?

It was installed and configured but not used by resolv.conf. Since my system is behind some measurements I have to adjust bind to forward all to our company dns - no direct outbound allowed here. As long the local bind wouldn't increase the cache timer I guess this wouldn't help much for cif.

>  i'd be interested if you have the same problem too by manually turning off the asn gatherer.. 

After disabling `asn.py*` no more timeout error.

**Comment by wesyoung** [comment]
ack. i guess maybe we need to make asn gatherers a little more configurable for places where outbound dns is restricted. @JustinAzoff's asn lookup stuff is promising for doing this sort of work "in-house" but not quite ready for prime time yet (feel free to fiddle with if you want).
